src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala

```diff
@@ -0,0 +1,224 @@
+    //Attempt to parse the expression into a pipeline query to hand off to Gremlin
+    //TODO: add caching here? Or move the creation of Gremlin pipelines here so we can cache those?
+    parseAll(path, expression) match {
+      case Success(result, _) => {
+        scala.util.Success(result)
+      }
+      case NoSuccess(msg, next) => {
+        scala.util.Failure(new RuntimeException("Failed at line %s, column %s: %s".format(next.pos.line, next.pos.column, msg)))
+      }
+    }
+  }
+
+  /** functions against pipes **/
+
+  // the root function starts the pipeline at some root entity type in the workspace
+  private def rootFunc(graphPipeline: PipeType, workspaceNamespace: WorkspaceNamespace, workspaceName: WorkspaceName, rootType: String, rootName: String): PipeResult = {
+    def workspaceFunc = new PipeFunction[Vertex, java.lang.Boolean] {
+      override def compute(a: Vertex) = {
+        propString(a, "_namespace") == workspaceNamespace && propString(a, "_name") == workspaceName
+      }
+    }
+
+    PipeResult(
+      // all vertexes of type Workspace filtered for the given namespace and name
+      graphPipeline.V("_clazz", classOf[Workspace].getSimpleName).filter(workspaceFunc)
+        // then entities from that workspace that match the root entity we are starting at
+        .out(rootType).filter(entityFunc(rootName)),
+
+      // text for what the pipeline looks like at this step
+      s"""new GremlinPipeline(graph).V("_entityType", "Workspace").filter(workspaceFunc).out($rootType,$rootName)"""
+    )
+  }
+
+  // get a vertex where the name matches the entity we are looking for
+  private def entityFunc(entityName: String) = new PipeFunction[Vertex, java.lang.Boolean] {
+    override def compute(a: Vertex) = {
+      propString(a, "_name") == entityName
+    }
+  }
+
+  // add pipe to an entity referenced by the current entity
+  private def entityNameAttributePipeFunc(entityRefName: String)(pipe: PipeType): PipeResult = {
+    PipeResult(
+      // an entity name is on the outgoing edge label
+      pipe.out(entityRefName),
+      // TODO: is there a way to automate returning this code-string from the code itself? (used for debugging purposes)
+      s".out($entityRefName)"
+    )
+  }
+
+  // TODO: We only support attributes here, not reference types (edges)
+  private def lastAttributePipeFunc(attributeName: String)(pipe: PipeType): FinalResult = {
+    val lastVertices = pipe.toList
+
+    if (lastVertices.size() == 0) {
+      throw new RuntimeException(s"Could not dereference $attributeName because pipe returned no entities")
+    }
+
+    // Look up the attributes on the list of vertices returned from the pipe. This will insert nulls into the list if
+    // the attribute doesn't exist.
+    FinalResult(lastVertices.map((v: Vertex) => {
+      v.getProperty(attributeName).asInstanceOf[String]
+    }), s".getProperty($attributeName)")
+  }
```

Comment:
In the future, we'll want to add the ability to return entities here if the last attribute is a reference.